icu4c/source/tools/toolutil/unewdata.cpp

```diff
@@ -56,6 +56,10 @@ udata_create(const char *dir, const char *type, const char *name,
         *pErrorCode=U_MEMORY_ALLOCATION_ERROR;
         return NULL;
     }
+
+    char dirSepChar = U_FILE_SEP_CHAR;
+    if(strchr(dir, U_FILE_SEP_CHAR) == NULL && strchr(dir, U_FILE_ALT_SEP_CHAR) != NULL)
```

Comment:
>>  It also looks like the change as-is fails to compile one several of the CI build bots.
> It still shows some errors. Any idea why?

It looks like the issue is the `gendict.exe` program is crashing, which causes the build to fail.

It seems the problem is that the input `dir` variable might be `NULL` in some cases. This causes the function `strchr` to deference a `NULL` pointer, causing the crash.

I'd probably suggest moving the check below inside the if block which already checks if `dir` is not `NULL`.

> With this PR I use heuristics that is already present in makeconv.cpp for all platforms.

Interesting... it seems that there is (for better or worse) already many places through the ICU code that does similar checks/hacks. 
Based on how this is done in other places in the ICU code, I'd likely suggest doing something like this instead:

```c++
    /* Check that the full path won't be too long */
    length = 0;					/* Start with nothing */
    if(dir != NULL  && *dir !=0)	/* Add directory length if one was given */
    {
#if (U_FILE_SEP_CHAR != U_FILE_ALT_SEP_CHAR)
        if(strchr(dir, U_FILE_SEP_CHAR) == NULL && strchr(dir, U_FILE_ALT_SEP_CHAR) != NULL)
        {
            dirSepChar = U_FILE_ALT_SEP_CHAR;
        }
#endif
```

(The use of the preprocessor macro will mean that the change only effects Windows/Cygwin platforms.)